Soc: Silabs: Optimize power management for Silabs Series 2

silabs-BastienB opened the pull request:
The objective of this pull request is to optimize the power of Silabs Series 2 devices.
This pull request does the following:
- Fix the implementation of the pm_state PM_STATE_SOFT_OFF to enter EM4.
- Remove the double call to WFI in pm_state_set() by removing the call to k_cpu_idle() and replacing it with its tracing and hook functions.
- Handle the PRIMASK correctly for sl_power_manager_sleep() by moving the setting of BASEPRI to 0.
- Retrieve the startup measurements correctly, reducing the early wakeup time.
- Remove the exit latency and minimum residency in Zephyr because the power manager already handles it.

asmellby commented on modules/hal_silabs/simplicity_sdk/config/sl_clock_manager_oscillator_config.h:
```diff
@@ -45,7 +45,7 @@
 	 : DT_PROP(DT_NODELABEL(hfrcodpll), clock_frequency) < 90000000 ? 80000000U                \
 									: 100000000U)
-#define SL_CLOCK_MANAGER_HFRCO_DPLL_EN DT_NUM_CLOCKS(DT_NODELABEL(hfrcodpll))
+#define SL_CLOCK_MANAGER_HFRCO_DPLL_EN DT_PROP(DT_NODELABEL(hfrcodpll), dpll_disable) ? 0U : 1U
```
`dpll-disable` isn't a valid property in the `silabs,series2-hfrcodpll` binding. If we want to do it this we, we'll need to also add it to the binding.

However, I'm not sure it's needed. The previous implementation checks for the presence of a parent clock to the HFRCODPLL, and if it has one, uses it as the DPLL reference clock. Disabling DPLL is therefore done by A) never defining a `clocks` property, or B) doing `/delete-property/ clocks;` if inheriting a DTS file that already set one. Having an explicit "disable" property seems to me to be duplicating that information. All SoC DTS files currently leave the HFRCODPLL free-running, and the board-level DTS sets the `clocks` property to enable DPLL if the board has a HFXO.

silabs-BastienB replied:
Alright, I can remove that commit if it's unnecessary. I was unaware of how to disable the DPLL without modifying the .dts directly.

jerome-pouiller commented on soc/silabs/common/soc_power_pmgr.c:
```diff
@@ -88,6 +74,32 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+{
+#if defined(HFXO_MANAGER_SLEEPTIMER_SYSRTC_INTEGRATION_ON)
+	sli_hfxo_manager_retrieve_begining_startup_measurement();
+#endif
```
Not related to this PR: I suggest to declare functions in a way it does not force `#ifdef` on the caller side.

The usual way, is to declare `sli_hfxo_manager_retrieve_begining_startup_measurement()` unconditionally but without definition. Thus the user get an error during the link. However, the user can use `if (IS_ENABLED(HFXO_MANAGER_SLEEPTIMER_SYSRTC_INTEGRATION_ON))` (note the current implementation of `IS_ENABLED()` expects the macro is defined to `1`).

An alternative is to use this pattern:
```
#ifdef HFXO_MANAGER_SLEEPTIMER_SYSRTC_INTEGRATION_ON
void sli_hfxo_manager_retrieve_begining_startup_measurement(void);
#else
static inline void sli_hfxo_manager_retrieve_begining_startup_measurement(void) { }
#endif
```

silabs-BastienB replied:
Unfortunately, a lot of functions in SiSDK are declared this way.

jerome-pouiller replied:
Yes, I know. This was a generic comment not related to your work.

jerome-pouiller commented on modules/hal_silabs/simplicity_sdk/config/sl_clock_manager_oscillator_config.h:
The commit title for `"modules/hal_silabs/simplicity_sdk/config: add an option to disable the DPLL of HFRCODPLL for Silabs Series 2 SoCs"` is too log. I suggest `"soc: silabs: Add an option to disable the DPLL of HFRCODPLL for Series-2"`

silabs-BastienB replied:
Noted, but I dropped this commit. See https://github.com/zephyrproject-rtos/zephyr/pull/84266#discussion_r1923383637

jerome-pouiller commented on dts/arm/silabs/efr32bg2x.dtsi:
```diff
@@ -130,9 +130,11 @@
+				min-residency-us = <0>;
 				/* HFXO remains active */
-				exit-latency-us = <2>;
+				exit-latency-us = <0>;
```
`0` is the default value for `min-residency-us` and `exit-latency-us` (see `PM_STATE_INFO_DT_INIT()`). You can drop these lines.

jerome-pouiller commented on dts/arm/silabs/efr32bg2x.dtsi:
```diff
@@ -142,8 +144,10 @@
+				/* The minimum residency and exit latency is
+				 * managed by sl_power_manager on S2 devices. */
+				min-residency-us = <0>;
+				exit-latency-us = <0>;
```
Ditto

jerome-pouiller commented on dts/arm/silabs/efr32bg2x.dtsi:
```diff
@@ -155,8 +159,10 @@
+				/* The minimum residency and exit latency is
+				 * managed by sl_power_manager on S2 devices. */
+				min-residency-us = <0>;
+				exit-latency-us = <0>;
```
and so on...

jerome-pouiller reviewed (commented):
Compliance job complains because your `user.email` is not correct. You can fix your `.gitconfig` and run `git rebase --exec 'git commit --amend --reset-author --no-edit' BASE`

silabs-BastienB committed:
soc: silabs: Fix EM4 enter for Silabs S2 SoCs

Fix PM_STATE_SOFT_OFF to properly enter EM4.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

silabs-BastienB committed:
soc: silabs: Fix double WFI for Silabs S2 SoCs

k_cpu_idle() and sl_power_manager_sleep() call WFI.
Remove the call to k_cpu_idle() and add back its tracing and
hook functions.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

silabs-BastienB committed:
soc: silabs: Fix the PRIMASK for Silabs S2 SoCs

sl_power_manager_sleep() doesn't expect the PRIMASK to be set when called.
Setting BASEPRI to 0 was moved to sl_power_manager_is_ok_to_sleep(),
this function is called after sl_power_manager_sleep() has set the PRIMASK.
Added sli_power_manager_on_wakeup() to force a clock restore before the
interrupt are handled. Added a call to retrieve the startup measurements,
reducing the early wakeup time.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

silabs-BastienB committed:
dts: arm: silabs: Remove exit latency and min residency on Silabs S2 SoCs

The exit latency and min residency is handled by sl_power_manager.
This improve power consumption by letting the device sleep longer.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

asmellby reviewed (dismissed):
Looks good to me other than the compliance warnings.

jerome-pouiller commented on soc/silabs/common/soc_power_pmgr.c:
```diff
@@ -63,9 +63,12 @@ void pm_state_set(enum pm_state state, uint8_t substate_id)
 	} else if (energy_mode != SL_POWER_MANAGER_EM0) {
+		/* Calling the tracing and hook functions provided in arch_cpu_idle(). */
+		IF_ENABLED(CONFIG_TRACING, (sys_trace_idle();))
+		IF_ENABLED(CONFIG_ARM_ON_ENTER_CPU_IDLE_PREPARE_HOOK, (z_arm_on_enter_cpu_idle_prepare();))
```
My bad, I made typo. I meant `IS_ENABLED()`:
```
	if (IS_ENABLED(CONFIG_TRACING)) {
		sys_trace_idle();
	}
	if (IS_ENABLED(CONFIG_ARM_ON_ENTER_CPU_IDLE_PREPARE_HOOK)) {
		z_arm_on_enter_cpu_idle_prepare();
	}
```

jhedberg commented:
@silabs-BastienB looks like this should be good enough now to be moved out from Draft state

asmellby commented on soc/silabs/common/soc_power_pmgr.c:
```diff
@@ -8,6 +8,7 @@
 #include <zephyr/logging/log.h>
 #include <zephyr/pm/pm.h>
 #include <sl_power_manager.h>
+#include <sli_hfxo_manager.h>
```
I get a compiler warning on xg24:
```
zephyr/soc/silabs/common/soc_power_pmgr.c:97:9: warning: implicit declaration of function 'sl_hfxo_manager_irq_handler'; did you mean 'sli_hfxo_manager_init_hardware'? [-Wimplicit-function-declaration]
   97 |         sl_hfxo_manager_irq_handler();
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         sli_hfxo_manager_init_hardware
```

The missing symbol seems to be in `sl_hfxo_manager.h`, so that should be included too.

asmellby commented on soc/silabs/common/soc_power_pmgr.c:
```diff
@@ -46,31 +47,25 @@ void pm_state_set(enum pm_state state, uint8_t substate_id)
+			sys_trace_idle();
+		}
+		if (IS_ENABLED(CONFIG_ARM_ON_ENTER_CPU_IDLE_PREPARE_HOOK)) {
+			z_arm_on_enter_cpu_idle_prepare();
```
`IS_ENABLED` doesn't work here, since the symbol definition is behind `#if defined(CONFIG_ARM_ON_ENTER_CPU_IDLE_PREPARE_HOOK)`:

```
zephyr/soc/silabs/common/soc_power_pmgr.c:60:25: warning: implicit declaration of function 'z_arm_on_enter_cpu_idle_prepare' [-Wimplicit-function-declaration]
   60 |                         z_arm_on_enter_cpu_idle_prepare();
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
